File: src/schemas/vaultwarden.ts
import { DecryptedToken } from '../decrypt';
import { SchemaFormatter } from './index';
export interface VaultwardenExport {
	items: VaultwardenItem[];
}

export interface VaultwardenItem {
	name: string;
	type: 1;
	login: {
		username: string;
		totp: string;
	};
}

export class VaultwardenFormatter implements SchemaFormatter {
	format(tokens: DecryptedToken[]): string {
		const bitwardenExport: VaultwardenExport = {
			items: [],
		};

		for (const token of tokens) {
			const name = token.name || '';
			const issuer = token.issuer || '';
			const secret = token.decrypted_seed || '';
			const digits = token.digits?.toString() || '6';

			const encodedIssuer = encodeURIComponent(issuer);
			const encodedName = encodeURIComponent(name);
			const encodedSecret = encodeURIComponent(secret);

			let totpUri = `otpauth://totp/${encodedIssuer}:${encodedName}?secret=${encodedSecret}&digits=${digits}`;
			if (issuer) {
				totpUri += `&issuer=${encodedIssuer}`;
			}

			bitwardenExport.items.push({
				name: name,
				type: 1,
				login: {
					username: name,
					totp: totpUri,
				},
			});
		}

		return JSON.stringify(bitwardenExport, null, 4);
	}
}

File: crates/core/src/schemas/vaultwarden.rs
use serde::Serialize;

use crate::decrypt::DecryptedToken;

use super::{encode_uri_component, SchemaFormatter};

#[derive(Serialize)]
struct VaultwardenExport {
    items: Vec<VaultwardenItem>,
}

#[derive(Serialize)]
struct VaultwardenItem {
    name: String,
    #[serde(rename = "type")]
    item_type: u8,
    login: VaultwardenLogin,
}

#[derive(Serialize)]
struct VaultwardenLogin {
    username: String,
    totp: String,
}

pub struct VaultwardenFormatter;

impl SchemaFormatter for VaultwardenFormatter {
    fn format(&self, tokens: &[DecryptedToken]) -> String {
        let mut export = VaultwardenExport { items: Vec::new() };

        for token in tokens {
            let name = token.name.clone();
            let issuer = token.issuer.clone().unwrap_or_default();
            let secret = token.decrypted_seed.clone();
            let digits = token.digits.to_string();

            let encoded_issuer = encode_uri_component(&issuer);
            let encoded_name = encode_uri_component(&name);
            let encoded_secret = encode_uri_component(&secret);

            let mut totp_uri = format!(
                "otpauth://totp/{encoded_issuer}:{encoded_name}?secret={encoded_secret}&digits={digits}"
            );
            if !issuer.is_empty() {
                totp_uri.push_str(&format!("&issuer={encoded_issuer}"));
            }

            export.items.push(VaultwardenItem {
                name: name.clone(),
                item_type: 1,
                login: VaultwardenLogin {
                    username: name,
                    totp: totp_uri,
                },
            });
        }

        super::to_string_pretty_indent(&export, b"    ")
    }
}
